paginate courses list page on backend

muhammad-ammar opened the pull request:
[EDUCATOR-882](https://openedx.atlassian.net/browse/EDUCATOR-882)

[Sandbox](https://discovery-pagination.sandbox.edx.org/publisher/courses/)

For oauth2 authentication, you can use the below credentials when asked
email: staff@example.com
password: edx

---

Below are some stats after pagination

**courses:** 1354
**user:** belongs to Publisher Admins group


|  time in seconds for 5 runs | Average Time(seconds)  |Records   |
|---|---|---|
| 1.27 + 1.22 + 0.397 +  0.390 + 0.402  | 0.735  |25   |
|  1.24 +  0.570 +  0.462 + 0.565  + 0.399 |0.647   |  50 |
|  1.74 +  0.848 +  0.791 +  0.716 +  0.787 | 0.976  |  100 |

muhammad-ammar commented:
@muzaffaryousaf @awais786 This is ready for review. Currently build is failing for some reason. Looking into it.

awais786 commented on course_discovery/templates/publisher/courses.html:
```diff
@@ -37,55 +36,92 @@ <h2 class="hd-2 course-count-heading">{{ object_list.count }} Courses</h2>
-            $('#dataTableCourse').DataTable({
-                "autoWidth": false,
-                "oLanguage": { "sEmptyTable": gettext("No courses have been created.") }
+            var $coursesTable = $('#dataTableCourse').DataTable({
```
Move this code in a separate file and call in a compress block.

muhammad-ammar replied:
ah..ok..thanks for pointing this.

muhammad-ammar replied:
I have created a separate js file but now how can i acces the server variables like `publisher_total_courses_count`, `courses`, `publisher_courses_url` etc in the js file? Can you please point me to any docs about why we need this compress change.

muzaffaryousaf commented:
@muhammad-ammar  Are you building up sandbox ?

muhammad-ammar commented:
@muzaffaryousaf I am trying to [build](https://docs.google.com/document/d/1pYMorQrXwBH_6B2dNn1ZmY84XDtX0QQ36GTJ_YgJoVg/edit?ts=596f673a#) a sandbox. Currently its not working.

muzaffaryousaf commented on course_discovery/apps/publisher/views.py:
```diff
@@ -817,15 +824,169 @@ def get_queryset(self):
+        ).data
+
+        if 'application/json' in request.META.get('HTTP_ACCEPT', ''):
+            totalRecords = self.object_list.count()
```
totalRecords --> count ?

muzaffaryousaf commented on course_discovery/static/js/publisher/views/courses.js:
```diff
@@ -0,0 +1,83 @@
+                "data": "course_title",
+                "render": function ( data, type, full, meta ) {
+                    if (data.url) {
+                        return '<a href="'+data.url+'">' + data.title + '</a>';
```
.format ?

muhammad-ammar replied:
I am not sure if we can use `format` in jS?

muzaffaryousaf replied:
NVM

muzaffaryousaf commented on course_discovery/apps/publisher/views.py:
```diff
@@ -817,15 +824,169 @@ def get_queryset(self):
+        if 'application/json' in request.META.get('HTTP_ACCEPT', ''):
+            totalRecords = self.object_list.count()
+            return JsonResponse({
+                'draw': int(self.request.GET['draw']),
```
What is this use for ?

muhammad-ammar replied:
this is required by DataTables. https://datatables.net/manual/server-side

muzaffaryousaf commented on course_discovery/apps/publisher/views.py:
```diff
@@ -817,15 +824,169 @@ def get_queryset(self):
+        ordering_field = ordering_fields.get(ordering_field_index)
+
+        if ordering_field == 'course_runs_count':
+            queryset = queryset.annotate(course_runs_count=Count('publisher_course_runs'))
```
Will this count related `publisher_course_runs` only?

muhammad-ammar replied:
yes

muzaffaryousaf commented on course_discovery/apps/publisher/tests/test_views.py:
```diff
@@ -157,7 +158,7 @@ def test_create_with_fail_transaction(self):
         data = {'number': 'course_2', 'image': make_image_file('test_banner.jpg')}
         course_dict = self._post_data(data, self.course)
-        with patch.object(Course, "save") as mock_method:
+        with mock.patch.object(Course, "save") as mock_method:
```
Were you seeing any failure for not using mock. or it's just refactor ?

muzaffaryousaf replied:
NVM, I've seen that you're using patch decorator.

muhammad-ammar replied:
I changed this due a quality error.

muhammad-ammar commented on course_discovery/apps/publisher/views.py:
```diff
@@ -817,15 +824,169 @@ def get_queryset(self):
+        except ValueError:
+            ordering_field_index = 0
+
+        ordering_direction = self.request.GET.get('sortDirection', 'asc')
```
This variable is passed from JS so JS convention is followed.

muzaffaryousaf replied:
I mean ordering_direction --> sort_direction

muzaffaryousaf replied:
If Django uses ordering then its okay

muhammad-ammar commented on course_discovery/apps/publisher/views.py:
```diff
@@ -817,15 +824,169 @@ def get_queryset(self):
+            ordering_field_index = 0
+
+        ordering_direction = self.request.GET.get('sortDirection', 'asc')
+        ordering_field = ordering_fields.get(ordering_field_index)
```
django uses the ordering convention so i followed that.

muzaffaryousaf replied:
ordering_field ??

muhammad-ammar commented on course_discovery/apps/publisher/views.py:
```diff
@@ -817,15 +824,169 @@ def get_queryset(self):
+        if ordering_direction == 'asc':
+            queryset = queryset.order_by(Lower(ordering_field).asc())
+        else:
+            queryset = queryset.order_by(Lower(ordering_field).desc())
```
for case insensitive ordering

muhammad-ammar commented on course_discovery/apps/publisher/views.py:
```diff
@@ -817,15 +824,169 @@ def get_queryset(self):
+
+        for keyword in keywords:
+            keyword_filter = Q(title__icontains=keyword) | Q(organizations__key__icontains=keyword)
+            keywords_filter = (keyword_filter & keywords_filter) if bool(keywords_filter) else keyword_filter
```
yes

muhammad-ammar commented on course_discovery/apps/publisher/views.py:
```diff
@@ -817,15 +824,169 @@ def get_queryset(self):
+
+        if dates:
+            query_filters.append(
+                Q(reduce(lambda x, y: x | y, [Q(course_state__owner_role_modified__date=date) for date in dates]))
```
yes...this will do an exact date match

muhammad-ammar commented on course_discovery/apps/publisher/serializers.py:
```diff
@@ -0,0 +1,109 @@
+        """
+        Returns count of course runs for a course.
+        """
+        return course.publisher_course_runs.count()
```
`publisher_course_runs` is a related field to course and will always be present.

muzaffaryousaf replied:
Any chance for does not exist?

muhammad-ammar commented on course_discovery/apps/publisher/views.py:
```diff
@@ -817,15 +824,169 @@ def get_queryset(self):
+            # 2: 'course_user_roles__user__full_name',
+            3: 'course_runs_count',
+            4: 'course_state__owner_role_modified',
+            5: 'course_state__owner_role_modified',
```
yes...both columns on frontend uses the same model field.

muzaffaryousaf replied:
And same data?

muhammad-ammar replied:
yes

muhammad-ammar commented on course_discovery/apps/publisher/views.py:
```diff
@@ -817,15 +824,169 @@ def get_queryset(self):
+        Get the number of items to paginate by.
+        """
+        try:
+            page_size = int(self.request.GET.get('length', COURSES_DEFAULT_PAGE_SIZE))
```
I will check if this is possible.

muhammad-ammar commented:
@muzaffaryousaf @awais786 Here is the [sandbox](https://discovery-pagination.sandbox.edx.org/publisher/courses/). Currently styling is messed up for some reason. Looking into it.

muzaffaryousaf commented:
@muhammad-ammar Sandbox is redirecting me to this url --> 
`http://127.0.0.1:8000/oauth2/authorize/?client_id=discovery-key&state=vp2K5UxMbANv5XFnm6WOb6qe4DvtHrrE&response_type=code&scope=openid+profile+email+permissions&redirect_uri=https://discovery-pagination.sandbox.edx.org/complete/edx-oidc/`

muhammad-ammar commented:
@muzaffaryousaf @awais786 Please continue the review. I have addressed the feedback and [sandbox](https://discovery-pagination.sandbox.edx.org/publisher/courses/) is working now. For oauth2 authentication, you can use the below credentials when asked

email: staff@example.com
password: edx

muhammad-ammar commented:
@sstack22 Can you take a look at pagination on [courses](https://discovery-pagination.sandbox.edx.org/publisher/courses/) list. Currently ordering is happening on 1st and last 3 columns and filtering is applying on first 2 and last 2 columns.

awais786 commented on course_discovery/apps/publisher/serializers.py:
```diff
@@ -0,0 +1,127 @@
+        """
+        Returns a dict containing course `title` and `url`.
+        """
+        publisher_hide_features_for_pilot = self.context['publisher_hide_features_for_pilot']
```
You can remove these kind of flags. They were added for pilot.

awais786 commented:
@muhammad-ammar I have checked the PR and sandbox also. Code look good and also working fine on sandbox. Nice work 👍

sstack22 commented:
@muhammad-ammar  - from looking at the sandbox, I can't tell if sorting/filtering is working as expected, as everything seems to have the same date (though the course names are changing order). Is there a bug here?

muhammad-ammar commented:
@sstack22 The date is same because all the courses data is imported from stage on same day. That's why same date is set for all the courses. I will manually modify the dates so that you can check the ordering. Will notify you once dates are updated.

muzaffaryousaf commented on course_discovery/apps/publisher/views.py:
```diff
@@ -817,15 +824,169 @@ def get_queryset(self):
+        context['publisher_courses_url'] = reverse('publisher:publisher_courses')
         return context
 
+    def get_paginate_by(self, queryset):
```
Do we need to address this ?

muhammad-ammar replied:
actually quality is good for this so that's why i didn't changed it.

muzaffaryousaf commented on course_discovery/apps/publisher/views.py:
```diff
@@ -817,15 +824,175 @@ def get_queryset(self):
+
+        if dates:
+            query_filters.append(
+                Q(reduce(lambda x, y: x | y, [
```
Not a big fan of use x and y as names. It'd be nice if we can change them.

muhammad-ammar replied:
how about `Q(reduce(lambda date1, date2: date1 | date2, [...`

muzaffaryousaf commented on course_discovery/apps/publisher/views.py:
```diff
@@ -817,15 +824,175 @@ def get_queryset(self):
+            filter_text (str): input text entered by user like 'intro to python 07/04/17'
+
+        Returns:
+            tuple: tuple of two lists, first list contains date strings and second contains keywords without dates
```
without dates ???

muhammad-ammar replied:
oops...looks like docstring need a cleanup

muzaffaryousaf commented on course_discovery/static/js/publisher/views/courses.js:
```diff
@@ -0,0 +1,88 @@
+        'autoWidth': false,
+        'processing': true,
+        'serverSide': true,
+        'lengthMenu': [25, 50, 100],
```
can we manage this from backend ?

muhammad-ammar commented:
@sstack22 Dates are changed now. You can check the ordering.

muhammad-ammar commented:
@muzaffaryousaf Addressed your feedback. Also sandbox is updated.

sstack22 commented:
👍  - this looks good!
